scripts/apogee/build_parent_sample.py

```diff
@@ -0,0 +1,283 @@
+import os
+import argparse
+import numpy as np
+from astropy.io import fits
+from astropy.table import Table, join, hstack
+from multiprocessing import Pool
+from tqdm import tqdm
+import healpy as hp
+import h5py
+import urllib
+
+# APOGEE shares a global wavelength grid
+lam = 10.0 ** np.arange(
+    4.179, 4.179 + 8575 * 6.0 * 10.0**-6.0, 6.0 * 10.0**-6.0
+).astype(np.float32)
+
+# detector edges
+blue_start = 246
+blue_end = 3274
+green_start = 3585
+green_end = 6080
+red_start = 6344
+red_end = 8335
+lam_cropped = lam[np.r_[blue_start:blue_end, green_start:green_end, red_start:red_end]]
+
+
+def download_allstar(base_path):
+    fullfoldername = os.path.join(base_path, "/spectro/aspcap/dr17/synspec_rev1/")
+    # Check if directory exists
+    if not os.path.exists(fullfoldername):
+        os.makedirs(fullfoldername)
+    filename = "allStar-dr17-synspec_rev1.fits"
+    fullfilename = os.path.join(fullfoldername, filename)
+    url = f"https://data.sdss.org/sas/dr17/apogee/spectro/aspcap/dr17/synspec_rev1/{filename}"
+    urllib.request.urlretrieve(url, fullfilename)
+
+
+def combined_spectra(
+        base_path,
+        field,
+        apogee,
+        telescope
+        ):
+    aspcap_code = "synspec_rev1"
+    str1 = f"https://data.sdss.org/sas/dr17/apogee/spectro/aspcap/dr17/{aspcap_code}/{telescope}/{field}/"
+
+    filename = f"aspcapStar-dr17-{apogee}.fits"
+    urlstr = str1 + filename
+
+    # check folder existence
+    fullfoldername = os.path.join(
+        base_path,
+        f"spectro/aspcap/dr17/{aspcap_code}/{telescope}",
+        str(f"{field}"),
+    )
+    if not os.path.exists(fullfoldername):
+        os.makedirs(fullfoldername)
+
+    fullfilename = os.path.join(fullfoldername, filename)
+
+    if os.path.exists(fullfilename):
+        return fullfilename
+    else:
+        try:
+            urllib.request.urlretrieve(urlstr, fullfilename)
+            return 0
+        except urllib.error.HTTPError as emsg:
+            return 1  # error code
+
+
+def visit_spectra(
+    base_path,
+    field,
+    telescope,
+    filename,
+):
+    str1 = f"https://data.sdss.org/sas/dr17/apogee/spectro/redux/dr17/stars/{telescope}/{field}/"
+    urlstr = str1 + filename
+
+    fullfoldername = os.path.join(
+        base_path,
+        f"spectro/redux/dr17/stars/{telescope}/",
+        str(f"{field}"),
+    )
+
+    if not os.path.exists(fullfoldername):
+        os.makedirs(fullfoldername)
+
+    fullfilename = os.path.join(fullfoldername, filename)
+    if os.path.exists(fullfilename):
+        return fullfilename
+    else:
+        try:
+            urllib.request.urlretrieve(urlstr, fullfilename)
+            return 0
+        except urllib.error.HTTPError as emsg:
+            return 1  # error code
+
+
+def selection_fn(base_path, catalog):
```

Comment:
can you move this selection function to the top of the file, to make it really easy for people to find?

Reply:
done